collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java

```diff
@@ -1476,99 +1467,77 @@ public boolean dispatchTouchEvent(MotionEvent mv) {
-            if (saveBeforeNextView(formController)) {
-                return;
-            }
-
-            int originalEvent = formController.getEvent();
-            int event = formController.stepToNextScreenEvent();
-
-            // Helps prevent transition animation at the end of the form (if user swipes left
-            // she will stay on the same screen)
-            if (originalEvent == event && originalEvent == FormEntryController.EVENT_END_OF_FORM) {
-                swipeHandler.setBeenSwiped(false);
-                return;
-            }
+    @Override
+    public void onSwipeForward() {
+        moveScreen(FORWARDS);
+    }
 
-            formIndexAnimationHandler.handle(getFormController().getFormIndex());
-        } catch (JavaRosaException e) {
-            Timber.d(e);
-            createErrorDialog(e.getCause().getMessage(), DO_NOT_EXIT);
-        }
+    @Override
+    public void onSwipeBackward() {
+        moveScreen(BACKWARDS);
     }
 
-    /**
-     * If moving backwards is allowed, displays the view for the previous question or field list.
-     * Steps the global {@link FormController} to the previous question and saves answers to the
-     * data model without checking constraints.
-     */
-    public void showPreviousView() {
-        if (allowMovingBackwards) {
-            state = null;
-            FormController formController = getFormController();
-            if (formController != null) {
-                // The answer is saved on a back swipe, but question constraints are ignored.
-                if (formController.currentPromptIsQuestion()) {
-                    saveAnswersForCurrentScreen(DO_NOT_EVALUATE_CONSTRAINTS);
-                }
+    private void moveScreen(Direction direction) {
+        FormController formController = getFormController();
+        if (formController == null) {
+            Timber.w("FormController has a null value");
+            swipeHandler.setBeenSwiped(false);
+            return;
+        }
 
-                try {
-                    if (formController.getEvent() != FormEntryController.EVENT_BEGINNING_OF_FORM) {
-                        int event = formController.stepToPreviousScreenEvent();
+        if (audioRecorderViewModel.isRecording().getValue()) {
+            // We want the user to stop recording before changing screens
+            DialogUtils.showIfNotShowing(RecordingWarningDialogFragment.class, getSupportFragmentManager());
+            swipeHandler.setBeenSwiped(false);
+            return;
+        }
 
-                        // If we are the beginning of the form, there is no previous view to show
-                        if (event == FormEntryController.EVENT_BEGINNING_OF_FORM) {
-                            event = formController.stepToNextScreenEvent();
-                            swipeHandler.setBeenSwiped(false);
+        if (direction == FORWARDS && formController.getEvent() == FormEntryController.EVENT_END_OF_FORM) {
```

Comment:
Why pull this direction-specific stuff out? Seems it should either be in the switch below or in the direction specific methods. Or if it's pulled out and their bodies are the same maybe it should be a single conditional.

If pulling it out, why not also pull the analogous beginning of form test out here? (Not that I think it's actually possible to be at the beginning of form event)

Reply:
I initially pulled all these out so the method had more of a "guard clauses followed by the happy logic" structure as it was really cluttered. That helped me work out what I could move out to the view models. I'll try putting them back into the `switch` now and see how it feels though.

Reply:
Ah it's good to keep them where they are so `state = null` happens in the same place. Initially that had to be in both parts of the `switch`. It would be cool to pair on this now you've got some context in your head. I'll ping you on Slack.